controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -52,42 +62,73 @@ boolean lock(Integer permit, long timeout, TimeUnit unit) throws InterruptedExce
         return result;
     }
 
-    void lockInterruptibly(Integer permit) throws InterruptedException {
+    boolean lockShared(final Integer permit, final long timeout, final TimeUnit unit) throws InterruptedException {
+        boolean result = false;
+        try {
+            result = lockSharedInterruptibly(permit, timeout, unit);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+        return result;
+    }
+
+    void lockInterruptibly(final Integer permit) throws InterruptedException {
         if (permit == null) {
             throw new IllegalArgumentException();
         }
         sync.acquireInterruptibly(permit);
     }
 
-    boolean lockInterruptibly(Integer permit, long timeout, TimeUnit unit) throws InterruptedException {
+    void lockSharedInterruptibly(final Integer permit) throws InterruptedException {
+        if (permit == null) {
+            throw new IllegalArgumentException();
+        }
+        sync.acquireSharedInterruptibly(permit);
+    }
+
+    boolean lockInterruptibly(final Integer permit, final long timeout, final TimeUnit unit) throws InterruptedException {
         if (permit == null) {
             throw new IllegalArgumentException();
         }
         return sync.tryAcquireNanos(permit, unit.toNanos(timeout));
     }
 
-    void unlock(Integer permit) {
+    boolean lockSharedInterruptibly(final Integer permit, final long timeout, final TimeUnit unit) throws InterruptedException {
+        if (permit == null) {
+            throw new IllegalArgumentException();
+        }
+        return sync.tryAcquireSharedNanos(permit, unit.toNanos(timeout));
+    }
+
+    void unlock(final Integer permit) {
         if (permit == null) {
             throw new IllegalArgumentException();
         }
         sync.release(permit);
     }
 
-    boolean detectDeadlockAndGetLock(int permit) {
+    void unlockShared(final Integer permit) {
+        if (permit == null) {
+            throw new IllegalArgumentException();
+        }
+        sync.releaseShared(permit);
+    }
+
+    boolean detectDeadlockAndGetLock(final int permit) {
         return sync.tryAcquire(permit);
     }
 
     private class Sync extends AbstractQueuedSynchronizer {
         private static final long serialVersionUID = 1L;
 
-        private final AtomicReference<Object> permitHolder = new AtomicReference<>(null);
+        private Integer permitHolder = null;
 
         @Override
-        protected synchronized boolean tryAcquire(int permit) {
+        protected boolean tryAcquire(final int permit) {
             if (compareAndSetState(0, 1)) {
-                permitHolder.set(permit);
+                permitHolder = permit;
                 return true;
-            } else if (permitHolder.get().equals(permit)) {
+            } else if (permitHolder != null && permitHolder.equals(permit)) {
```

Comment:
The permitHolder.equals(permit) call can NPE if tryRelease has been called concurrently and permitHolder was set to null after the null check.

The existing code would have had this problem too, but the methods being synchronized prevented that.

At the cost of an object allocation, this is safe:

```
} else if (Integer.valueOf(permit).equals(permitHolder)) {
```